openmeter/productcatalog/subscription/http/change.go

```diff
@@ -0,0 +1,133 @@
+package httpdriver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/openmeterio/openmeter/api"
+	planhttp "github.com/openmeterio/openmeter/openmeter/productcatalog/plan/httpdriver"
+	plansubscription "github.com/openmeterio/openmeter/openmeter/productcatalog/subscription"
+	"github.com/openmeterio/openmeter/openmeter/subscription"
+	"github.com/openmeterio/openmeter/pkg/convert"
+	"github.com/openmeterio/openmeter/pkg/framework/commonhttp"
+	"github.com/openmeterio/openmeter/pkg/framework/transport/httptransport"
+	"github.com/openmeterio/openmeter/pkg/models"
+)
+
+type (
+	ChangeSubscriptionRequest  = plansubscription.ChangeSubscriptionRequest
+	ChangeSubscriptionResponse = api.SubscriptionChangeResponseBody
+	ChangeSubscriptionParams   = struct {
+		ID string
+	}
+	ChangeSubscriptionHandler = httptransport.HandlerWithArgs[ChangeSubscriptionRequest, ChangeSubscriptionResponse, ChangeSubscriptionParams]
+)
+
+func (h *handler) ChangeSubscription() ChangeSubscriptionHandler {
+	return httptransport.NewHandlerWithArgs(
+		func(ctx context.Context, r *http.Request, params ChangeSubscriptionParams) (ChangeSubscriptionRequest, error) {
+			var body api.ChangeSubscriptionJSONRequestBody
+
+			if err := commonhttp.JSONRequestBodyDecoder(r, &body); err != nil {
+				return ChangeSubscriptionRequest{}, err
+			}
+
+			ns, err := h.resolveNamespace(ctx)
+			if err != nil {
+				return ChangeSubscriptionRequest{}, err
+			}
+
+			// Any transformation function generated by the API will succeed if the body is serializable, so we have to check for the presence of
+			// fields to determine what body type we're dealing with
+			type testForCustomPlan struct {
+				CustomPlan any `json:"customPlan"`
+			}
+
+			var t testForCustomPlan
+
+			bodyBytes, err := json.Marshal(body)
```

Comment:
why do we need to double parse body?

Reply:
To be able to determine which request body we received. It's commented just above, but the body.AsXY functions will always succeed if the body is deserializable (there isn't a discriminator for this type that we could check for). We can do it without the extra parsing, it was just less awkward to do it like this, as the values body.AsXY returns cannot easily be checked for being empty (they aren't pointers but structs)